examples/moe_dualpipe/examples/moe_train_basic.py

```diff
@@ -0,0 +1,288 @@
+from typing import List, Optional, Callable, Tuple
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as dist
+import os
+
+from dualpipe import DualPipe, set_p2p_tensor_shapes, set_p2p_tensor_dtype
+from dualpipe.utils import WeightGradStore, run_backward
+
+class MoEConfig:
+    def __init__(
+        self,
+        hidden_size: int,
+        num_experts: int = 4,
+        top_k: int = 2,
+        aux_loss_coeff: float = 0.01,
+        z_loss_coeff: float = 0.0001
+    ):
+        self.hidden_size = hidden_size
+        self.intermediate_size = hidden_size * 4
+        self.num_experts = num_experts
+        self.top_k = top_k
+        self.moe_aux_loss_coeff = aux_loss_coeff
+        self.moe_z_loss_coeff = z_loss_coeff
+        self.initializer_range = 0.02
+
+class MoEAuxLossAutoScaler(torch.autograd.Function):
```

Comment:
If these MOE model definition codes are copied/modified from other repo's codes, add comments stating the original code source.